### PR TITLE
Added copy-api service implementation

### PR DESCRIPTION
**Pull request by thilinicooray**
-Renamed API property "swagger" to apiDefinition
-Enabled POST and PUT for api object with swagger
-Added location header for POST responses

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java [hunk 1]
```diff
@@ -89,18 +92,24 @@ public Response apisPost(APIDTO body,String contentType){
                 PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
             }
             apiProvider.addAPI(apiToAdd);
-
+            apiProvider.saveSwagger20Definition(apiToAdd.getId(), body.getApiDefinition());
+            //This was added in order to provide location header for HTTP response
+            APIIdentifier createdApiId = apiToAdd.getId();
+            createdApiUri = new URI("http://10.100.7.39:9763/api/am/v1/apis/"
```
Comment (abimarank) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Why is this URL hard coded?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java [hunk 2]
```diff
@@ -109,22 +118,56 @@ public Response apisChangeLifecyclePost(String newState,String publishToGateway,
...
+        URI newVersionedApiUri = null;
+        APIDTO newVersionedApi = null;
+
+        try {
+            APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifier(apiId);
+            String loggedInUser = CarbonContext.getThreadLocalCarbonContext().getUsername();
+            APIProvider apiProvider = RestApiUtil.getProvider(loggedInUser);
+            String tenantDomain =  CarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+            if(tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                isTenantFlowStarted = true;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Move the flag

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java [hunk 3]
```diff
@@ -109,22 +118,56 @@ public Response apisChangeLifecyclePost(String newState,String publishToGateway,
...
+            org.wso2.carbon.apimgt.api.model.API api = apiProvider.getAPI(apiIdentifier);
+            if (api != null) {
+                apiProvider.createNewAPIVersion(api, newVersion);
+                //get newly created API to return as response
+                APIIdentifier apiNewVersionedIdentifier =
+                    new APIIdentifier(apiIdentifier.getProviderName(), apiIdentifier.getApiName(), newVersion);
+                newVersionedApi = APIMappingUtil.fromAPItoDTO(apiProvider.getAPI(apiNewVersionedIdentifier));
+                //This was added in order to provide location header for HTTP response
+                newVersionedApiUri =
+                    new URI("http://10.100.7.39:9763/api/am/v1/apis/" +
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: hardcoded urls

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java (hunk 1 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Fixed in PR #1077 commit : f0a44cc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java (hunk 3 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Fixed in PR #1077 commit : f0a44cc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java (hunk 2 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: The tenants flows have been moved currently from impl level to CXF handler level
Fixed comments in #1091 commit : 897ef52

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java [hunk 4]
```diff
@@ -109,22 +118,56 @@ public Response apisChangeLifecyclePost(String newState,String publishToGateway,
...
+                        apiIdentifier.getApiName() + "-" + newVersion + "-" + apiIdentifier.getProviderName());
+            } else {
+                throw new NotFoundException();
+            }
+
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Wrong exception handling. Please fix.

Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Fixed in PR #1280

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/gen/java/org/wso2/carbon/apimgt/rest/api/dto/APIDTO.java [hunk 5]
```diff
@@ -306,7 +307,7 @@ public String toString()  {
     sb.append("  context: ").append(context).append("\n");
     sb.append("  version: ").append(version).append("\n");
     sb.append("  provider: ").append(provider).append("\n");
-    sb.append("  swagger: ").append(swagger).append("\n");
+    sb.append("  apiDefinition: ").append(apiDefinition).append("\n");
     sb.append("  status: ").append(status).append("\n");
     sb.append("  responseCaching: ").append(responseCaching).append("\n");
     sb.append("  cacheTimeout: ").append(cacheTimeout).append("\n");
```
Comment (malakasilva) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/gen/java/org/wso2/carbon/apimgt/rest/api/dto/APIDTO.java]: Better if we can include the \n in a constants file

Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/gen/java/org/wso2/carbon/apimgt/rest/api/dto/APIDTO.java]: Please note that all the files in components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher/src/gen including DTO classes get auto-generated using swagger-to-cxf plugin based on the modifications done for api.yaml

**Pull request merged**